File: 0885-spiral-matrix-iii/0885-spiral-matrix-iii.cs
public class Solution {
    public int[][] SpiralMatrixIII(int rows, int cols, int rStart, int cStart) {
        int[][] directions = new int[][]{
            new int[] {0,1},
            new int[] {1,0},
            new int[] {0,-1},
            new int[] {-1,0}
        };
        
        List<int[]> result = new List<int[]>();
        result.Add(new int[] {rStart, cStart});
        
        int steps = 1;
        int d = 0;
        
        while ( result.Count < rows * cols){
            for (int i = 0; i < 2; i++){
                for (int j = 0; j < steps; j++){
                    rStart += directions[d][0];
                    cStart += directions[d][1];
                    if ( 0 <= rStart && rStart < rows && 0 <= cStart && cStart < cols){
                        result.Add(new int[] {rStart, cStart });
                    }
                }
                d = (d + 1) % 4;
            }
            steps ++;
        }
        return result.ToArray();
    }
}

File: 0885-spiral-matrix-iii/0885-spiral-matrix-iii.py
class Solution:
    def spiralMatrixIII(self, rows: int, cols: int, rStart: int, cStart: int) -> List[List[int]]:

        ans = [(rStart,cStart)]
        if rows * cols ==1:
            return ans

        # for walk length k = 1,3,5,7...

        for k in range(1,2*(rows+cols),2):

            for dr,dc,dk in ((0,1,k),(1,0,k),(0,-1,k+1),(-1,0,k+1)):

                for _ in range(dk): # walk in this direction for dk times

                    rStart += dr
                    cStart += dc

                    if 0 <=rStart < rows and 0 <= cStart <cols:
                        ans.append((rStart,cStart))
                        if len(ans) == rows * cols:
                            return ans
        